Improve recover command

aawsome opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Enhance the `recover` command to make it useful in more recovery situations:

- only save data not referenced by any snapshot
- dont't write an empty snapshot
- use progress bar for scanning trees

OLD AND REMOVED:

> 
> - allow to combine all found directories in one snapshot (equals the current behavior; default) or make one snapshot per "hanging" root tree
> - allow to add user-defined tags to newly generated snapshots
> - add dry-run possibility
> 
> Especially the flags `--split=true` seem very useful for me:  If you `forget` a couple of snapshots and run `restic recover --split=true`, you'll get the removed snapshots back (but deduplicated in the sense that multiple completely unchanged snapshots will resurrect only once and without snapshot metadata).
> 

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

no.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- I have not added tests for all changes in this PR
- I have not added documentation for the changes (in the manual) - there is no docu for `recover` so far
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

aawsome commented:
rebased this to resolve the merge conflicts.
I have to add a changelog entry and maybe docu for the new recover command (there is no docu for the current `recover`, though), but besides that, this PR is ready for review.

MichaelEischer commented on cmd/restic/cmd_recover.go:
```diff
@@ -24,15 +25,29 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
 	cmdRoot.AddCommand(cmdRecover)
+	recoverFlags := cmdRecover.Flags()
+	recoverFlags.BoolVar(&recoverOptions.All, "all", false, "add all directories, not only unreferenced ones")
+	recoverFlags.BoolVar(&recoverOptions.Combine, "combine", true, "if true, combine all directories into one snapshot, else generate multple snapshots")
```
With the new prune implementation I'd expect there to be thousands of unreferenced trees, so I doubt that this option would still be useful. A option which must be passed as `--combine=false` is really confusing.

aawsome replied:
Thanks for the hint about the pruning. This applies to `recover` with and without combine. The only difference is whether the unreferenced trees are combined in one new snapshot (as current `recover` does) or in subtrees within one single new snapshot. I would always prefer to use `--combine=false` as this allows to clean up the unneeded snapshots.
But as you can specify tags for the "uncombined" snapshots, even lots of snapshots could be handled or easily removed if not needed.

About the option name: What about inverting it: `--split` with default "false"?

aawsome replied:
About the unreferenced trees: I made #3228 and think we should merge this in order to save the `recover` command (independent from this PR).

I already changed the name to `--split`.

MichaelEischer commented on cmd/restic/cmd_recover.go:
```diff
@@ -24,15 +25,29 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
 func init() {
 	cmdRoot.AddCommand(cmdRecover)
+	recoverFlags := cmdRecover.Flags()
+	recoverFlags.BoolVar(&recoverOptions.All, "all", false, "add all directories, not only unreferenced ones")
```
Do you see a use case where one would want to add root directories already covered by existing snapshots?

aawsome replied:
No, actually not :wink: 
The current version effectively is `--all=true` and therefore I added this option to preserve the current behavior. I can remove this option and always remove directories covered by existing snapshots.

aawsome replied:
I removed the `--all` option.

MichaelEischer commented on cmd/restic/cmd_recover.go:
```diff
@@ -24,15 +25,29 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
+	recoverFlags := cmdRecover.Flags()
+	recoverFlags.BoolVar(&recoverOptions.All, "all", false, "add all directories, not only unreferenced ones")
+	recoverFlags.BoolVar(&recoverOptions.Combine, "combine", true, "if true, combine all directories into one snapshot, else generate multple snapshots")
+	recoverFlags.BoolVarP(&recoverOptions.DryRun, "dry-run", "n", false, "only show what would be done")
```
`--dry-run` only seems to be useful in combination with `--combine=false`?

aawsome replied:
Another use case would be to use it for debugging/troubleshooting which would also work on read-only repositories.

MichaelEischer commented on cmd/restic/cmd_recover.go:
```diff
@@ -59,90 +74,113 @@ func runRecover(gopts GlobalOptions) error {
-			trees[subtree] = true
+	if !opts.All {
+		Verbosef("load snapshots\n")
+		snapshots, err := restic.LoadAllSnapshots(gopts.ctx, repo, nil)
```
That should now use `ForAllSnapshots`.

aawsome replied:
is changed.

MichaelEischer commented on cmd/restic/cmd_recover.go:
```diff
@@ -59,90 +74,113 @@ func runRecover(gopts GlobalOptions) error {
-		if seen {
-			continue
+		if !seen {
+			Verbosef("found root tree %v\n", id.Str())
```
`s/Verbosef/Verboseff`. Printing all root tree ids is too much for normal verbosity.

aawsome replied:
changed.

MichaelEischer reviewed (changes requested):
I don't think that the `recover` command needs that many flags. For me it looks like `--all=false` is the behavior everyone would. `--all=true` just clutters the recovered snapshot(s). Now with the new prune implementation, there will be hundred or thousands of "root" trees in the repository, which makes `--combine=false` essentially useless in my opinion. When `recover` only creates a single snapshot, I don't see much use for a `--dry-run` option either, in the worst case `recover` will create a single snapshot which is also tagged accordingly.

`--all=false --combine=false` will not really recover deleted snapshots, these still miss all metadata.

aawsome commented on cmd/restic/cmd_recover.go:
```diff
@@ -24,15 +25,29 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
+	All     bool
+	Combine bool
+	DryRun  bool
+	Tags    []string
```
is changed.

rawtaz commented:
What's the use case of tagging a snapshot created by `recover`? This command is something you'd run manually when needed, so  why would you need to tag it? Even if you do, you could just use the existing `tag` command so there's little need for adding code to tag these manually produced snapshots IMO.

aawsome commented:
> What's the use case of tagging a snapshot created by `recover`? This command is something you'd run manually when needed, so why would you need to tag it? Even if you do, you could just use the existing `tag` command so there's little need for adding code to tag these manually produced snapshots IMO.

With `--split` you can get quite a lot of new snapshots. I thought it might be handy to let `recover` tag them all.
And it is exactly 3 extra lines of code... 

> Also, I don't understand the point of having a boolean argument to `--split`. If the default is to not split, then just `--split` would be enough, no?

You are right. Of course `--split` also works.

aawsome committed:
Improve recover

- only save directories not referenced by any snapshot
- dont't write empty snapshot
- use progress bar

aawsome commented:
rebased to solve merge conflicts.

MichaelEischer reviewed (changes requested):
I'm still not convinced about the `--split` option as it just feels like a crutch to me. After all, the whole metadata of a snapshot is still lost in the process. So, I'd like to limit this PR to the code cleanups and tweaks, but defer new options to a discussion on how "un-deleting" a snapshot could/should work.

Please remove the `--split` and the then obsolete `--dry-run` and `--tags` options. The factored out `createSnapshot` function can stay though.

aawsome commented on cmd/restic/cmd_recover.go:
```diff
@@ -59,90 +73,111 @@ func runRecover(gopts GlobalOptions) error {
 		for _, node := range tree.Nodes {
-			if node.Type != "dir" || node.Subtree == nil {
-				continue
+			if node.Type == "dir" && node.Subtree != nil {
```
Very true :+1: 
Let's see when I'll find some time for it :wink:

aawsome commented:
I'm fine with postponing the '--split' or something else to a future PR.

> Please remove the `--split` and the then obsolete `--dry-run` and `--tags` options. The factored out `createSnapshot` function can stay though.

Is removed. Hope these small improvements are nevertheless helpful!

MichaelEischer reviewed (approved):
LGTM.

Excluding trees belonging to snapshot might be quite helpful, especially for a repository containing lots of snapshots. And the code cleanup is also a plus :-) .
